safehttp/response_writer.go

```diff
@@ -64,21 +66,68 @@ func NotWritten() Result {
 	return Result{}
 }
 
-// Write TODO
+// Write dispatches the response to the Dispatcher, setting the Content-Type to
+// text/html; charset=utf-8 and response status to 200 OK if resp is a safe
+// response. The Dispatcher will then write the response to the underlying
+// Response Writer.
+//
+// TODO: replace panics with proper error handling when getting the response
+// Content-Type or writing the response fails.
 func (w *ResponseWriter) Write(resp Response) Result {
-	w.markWritten()
+	ct, err := w.d.ContentType(resp)
+	if err != nil {
+		panic(err)
+	}
+	w.rw.WriteHeader(int(StatusOK))
+	w.rw.Header().Set("Content-Type", ct)
```

Comment:
You need to set the ct header before calling `WriteHeader`. The headers of the response are written when you call `WriteHeader`, so they must be modified first.

The underlying responsewriter will write a 200 OK response if `WriteHeader` was never called, so you can remove the call to `WriteHeader`. And if there is a error during `w.d.Write` which results in a panic then something else will try to write to the responsewriter and in that case it would be bad if the response code is already set.

So remove the call to `WriteHeader` here, in `WriteJSON` and in `WriteTemplate`.

Reply:
1. I agree that the headers need to be set before doing `WriteHeader`.
2. I'd still explicitly call `WriteHeader(200)` here, though.

Reply:
Imagine that `w.d.Write` panics or returns an error, in which case `ResponseWriter.Write` will panic. This panic will be recovered by the framework (ServeMux) and a 500 Internal Server Error response will be written to the `ResponseWriter`. But the status code and the headers have already been written, so things will break.

We need to remove the call to `WriteHeader`.

Reply:
Right, we discussed this offline yesterday. We need a better way of handling errors instead of the panics. Right now, if the Dispatcher errs after writing half of the response, the status code and content type had already been set and there's not much we can do as we cannot reset it, right? An option is to buffer the response and only send it after we make sure everything works as intended, but IIRC that's a feature that we don't plan to introduce in V0.

Also, BTW, these are problems that don't come up in tests because we use the `responseRecorder`. This is not ideal IMO as it's easy to introduce bugs (e.g. my bad ordering of statements) and tests won't fail. Should we do something about it?

Reply:
I am not sure what to do here.

Reply:
I agree that we need a better error strategy for the dispatcher, but as you said, that's not V0.

To detect these kinds of errors in tests we need thorough integration tests between the ServeMux, the ResponseWriter and a Dispatcher.

But I still think that the ResponseWriter should not call `WriteHeader`. The Dispatcher should start doing the actual writing to the network connection.

Reply:
I am not sure I understand the last sentence.